Do not Panic so easy

When executing a check of a MerkleProof generated by an old version of this repository, we experience a Panic. 

Instead of Panic I think an Error should be returned, specially if the main aim of this repository is to be used as an external library by other projects.

```
020-03-18T18:42:09Z	INFO	vochain/app.go:47	tendermint Core version: 0.32.8
2020-03-18T18:42:09Z	INFO	vochain/app.go:48	tendermint P2P protocol version: 7
2020-03-18T18:42:09Z	INFO	vochain/app.go:49	tendermint Block protocol version: 10
2020-03-18T18:42:09Z	INFO	vochain/app.go:52	current height is 223815, current APP hash is 75894b2e73cb5a03615fc4d16be8d008394407a1318b48a7130cb9104172506c
2020-03-18T18:42:09Z	DEBUG	vochain/apputils.go:218	generated nullifier: 08473c66351c2dbbc726effed933a56fc03a0f76b715011f82aa07d8d842bd2e
2020-03-18T18:42:09Z	DEBUG	vochain/apputils.go:227	extracted pubkey: 041423328db7ba2f02e00dca1fcca05509c5aa14ba0e6227c1b5097c3b094b40d5d59208ba804f1ea9a07a9f261ec8fc7d4c982ad64c9c7f57979dc014d2e646f5
panic: inputs values not inside Finite Field

goroutine 35348 [running]:
github.com/iden3/go-iden3-core/merkletree.HashElems(0xc001929f38, 0x2, 0x2, 0xc011e34580)
	/go/pkg/mod/github.com/iden3/go-iden3-core@v0.0.7-0.20200306161755-b1ca8370e595/merkletree/utils.go:72 +0xe2
github.com/iden3/go-iden3-core/merkletree.(*Node).Key(0xc00192a008, 0xc012bf3ca0)
	/go/pkg/mod/github.com/iden3/go-iden3-core@v0.0.7-0.20200306161755-b1ca8370e595/merkletree/node.go:99 +0x26b
github.com/iden3/go-iden3-core/merkletree.RootFromProof(0xc005567680, 0xc012bf3ca0, 0xc012bf3fe0, 0x0, 0xc011e34500, 0xc012bf3fe0)
	/go/pkg/mod/github.com/iden3/go-iden3-core@v0.0.7-0.20200306161755-b1ca8370e595/merkletree/merkletree.go:901 +0x202
github.com/iden3/go-iden3-core/merkletree.VerifyProof(0xc00192a100, 0xc005567680, 0xc012bf3ca0, 0xc012bf3fe0, 0xeb490ba9e585e996)
	/go/pkg/mod/github.com/iden3/go-iden3-core@v0.0.7-0.20200306161755-b1ca8370e595/merkletree/merkletree.go:864 +0x3f
gitlab.com/vocdoni/go-dvote/tree.CheckProof(0xc014415940, 0x40, 0xc001020480, 0x102, 0xc005f64fa0, 0x78, 0xa0, 0x0, 0x0, 0x0)
	/src/tree/tree.go:128 +0x415
gitlab.com/vocdoni/go-dvote/vochain.checkMerkleProof(...)
	/src/vochain/apputils.go:345
gitlab.com/vocdoni/go-dvote/vochain.VoteTxCheck(0xc011d10f08, 0x6, 0x0, 0x0, 0xc0056741e0, 0x42, 0xc001020480, 0x102, 0xc00687c7e0, 0x82, ...)
	/src/vochain/apputils.go:232 +0x847
gitlab.com/vocdoni/go-dvote/vochain.ValidateTx(0xc00016b180, 0x265, 0x265, 0xc00145eb10, 0x4, 0x100000000, 0x0, 0xc0011ebb01)
	/src/vochain/apputils.go:47 +0x33f
gitlab.com/vocdoni/go-dvote/vochain.ValidateAndDeliverTx(0xc00016b180, 0x265, 0x265, 0xc00145eb10, 0xc015bd599b, 0x3, 0x36a49, 0x1ecaadfa, 0xed5f86a2d)
	/src/vochain/apputils.go:74 +0x6a
gitlab.com/vocdoni/go-dvote/vochain.(*BaseApplication).DeliverTx(0xc029670660, 0xc00016b180, 0x265, 0x265, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/src/vochain/app.go:121 +0x7e
github.com/tendermint/tendermint/abci/client.(*localClient).DeliverTxAsync(0xc027a06f00, 0xc00016b180, 0x265, 0x265, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/pkg/mod/github.com/tendermint/tendermint@v0.32.8/abci/client/local_client.go:88 +0xdc
github.com/tendermint/tendermint/proxy.(*appConnConsensus).DeliverTxAsync(0xc005210d70, 0xc00016b180, 0x265, 0x265, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/pkg/mod/github.com/tendermint/tendermint@v0.32.8/proxy/app_conn.go:73 +0x5e
github.com/tendermint/tendermint/state.execBlockOnProxyApp(0x2fbb440, 0xc0145e0660, 0x2fcc380, 0xc005210d70, 0xc00000d0e0, 0x2fddc20, 0xc02811c128, 0x6, 0xc0145de1c9, 0x3)
	/go/pkg/mod/github.com/tendermint/tendermint@v0.32.8/state/execution.go:293 +0x610
github.com/tendermint/tendermint/state.(*BlockExecutor).ApplyBlock(0xc011c9f490, 0xa, 0x0, 0xc0145de1c0, 0x6, 0xc0145de1c9, 0x3, 0x36a48, 0x45, 0xc0297ccc00, ...)
	/go/pkg/mod/github.com/tendermint/tendermint@v0.32.8/state/execution.go:131 +0x17a
github.com/tendermint/tendermint/consensus.(*Handshaker).replayBlock(0xc00192d318, 0xa, 0x0, 0xc0145de1c0, 0x6, 0xc0145de1c9, 0x3, 0x36a48, 0x45, 0xc0297ccc00, ...)
	/go/pkg/mod/github.com/tendermint/tendermint@v0.32.8/consensus/replay.go:475 +0x233
github.com/tendermint/tendermint/consensus.(*Handshaker).replayBlocks(0xc00192d318, 0xa, 0x0, 0xc0145de1c0, 0x6, 0xc0145de1c9, 0x3, 0x36a48, 0x45, 0xc0297ccc00, ...)
	/go/pkg/mod/github.com/tendermint/tendermint@v0.32.8/consensus/replay.go:455 +0x3f3
github.com/tendermint/tendermint/consensus.(*Handshaker).ReplayBlocks(0xc002a3f318, 0xa, 0x0, 0xc0145de1c0, 0x6, 0xc0145de1c9, 0x3, 0x36a48, 0x45, 0xc0297ccc00, ...)
	/go/pkg/mod/github.com/tendermint/tendermint@v0.32.8/consensus/replay.go:386 +0xedf
github.com/tendermint/tendermint/consensus.(*Handshaker).Handshake(0xc00192d318, 0x2fdf6a0, 0xc0297c6a10, 0x38, 0x8)
	/go/pkg/mod/github.com/tendermint/tendermint@v0.32.8/consensus/replay.go:269 +0x485
github.com/tendermint/tendermint/node.doHandshake(0x2fddc20, 0xc02811c128, 0xa, 0x0, 0xc0145de1c0, 0x6, 0xc0145de1c9, 0x3, 0x36a48, 0x45, ...)
	/go/pkg/mod/github.com/tendermint/tendermint@v0.32.8/node/node.go:281 +0x19a
github.com/tendermint/tendermint/node.NewNode(0xc00211a280, 0x2fac5c0, 0xc0066c4000, 0xc000c71ef0, 0x2f5ffe0, 0xc029670c40, 0xc000cb8850, 0x2c6bd80, 0xc000cb8890, 0x2fbb440, ...)
	/go/pkg/mod/github.com/tendermint/tendermint@v0.32.8/node/node.go:596 +0x343
gitlab.com/vocdoni/go-dvote/vochain.newTendermint(0xc029670660, 0xc0011a0000, 0xc002346900, 0x828, 0x828, 0xc0066c4000, 0x2c6d648, 0x0, 0xed4f76)
	/src/vochain/start.go:223 +0xb6a
gitlab.com/vocdoni/go-dvote/vochain.NewVochain.func1(0xc029670660, 0xc0011a0000, 0xc002346900, 0x828, 0x828, 0x0, 0xc000e94dd0)
	/src/vochain/start.go:43 +0x73
created by gitlab.com/vocdoni/go-dvote/vochain.NewVochain
	/src/vochain/start.go:42 +0x20c
```

--- a/components/verifier/verifier.go
+++ b/components/verifier/verifier.go
@@ -47,7 +47,11 @@ func (v *Verifier) VerifyCredentialExistence(credExist *proof.CredentialExistenc
 	}
 	// Verify that the idenState is built from claims merkle tree where the
 	// claim exists.
-	claimsRoot, err := merkletree.RootFromProof(credExist.MtpClaim, credExist.Claim.HIndex(), credExist.Claim.HValue())
+	hi, hv, err := credExist.Claim.HiHv()
+	if err != nil {
+		return err
+	}
+	claimsRoot, err := merkletree.RootFromProof(credExist.MtpClaim, hi, hv)
 	if err != nil {
 		return err
 	}
@@ -107,7 +111,11 @@ func (v *Verifier) VerifyCredentialValidity(credValid *proof.CredentialValidity,
 	// NOTE: Once we add versions, this will require some changes that need to be thought properly!
 	nonce := claims.GetRevocationNonce(credValid.CredentialExistence.Claim)
 	revLeaf := claims.NewLeafRevocationsTree(nonce, 0xffffffff).Entry()
-	revocationsTreeRoot, err := merkletree.RootFromProof(credValid.MtpNotNonce, revLeaf.HIndex(), revLeaf.HValue())
+	hi, hv, err := revLeaf.HiHv()
+	if err != nil {
+		return err
+	}
+	revocationsTreeRoot, err := merkletree.RootFromProof(credValid.MtpNotNonce, hi, hv)
 	if err != nil {
 		return err
 	}

--- a/components/verifier/verifier_test.go
+++ b/components/verifier/verifier_test.go
@@ -108,7 +108,7 @@ func TestVerifyCredentialExistence(t *testing.T) {
 	// Cred Exist has bad RootsTreeRoot
 	credExistBad = &proof.CredentialExistence{}
 	Copy(credExistBad, credExist)
-	credExistBad.RootsTreeRoot[0] = 0x00
+	credExistBad.RootsTreeRoot[1] ^= 0xff
 	require.NotEqual(t, credExist, credExistBad)
 	err = verifier.VerifyCredentialExistence(credExistBad)
 	assert.NotNil(t, err)
@@ -116,7 +116,7 @@ func TestVerifyCredentialExistence(t *testing.T) {
 	// Cred Exist has bad IdenState
 	credExistBad = &proof.CredentialExistence{}
 	Copy(credExistBad, credExist)
-	credExistBad.IdenStateData.IdenState[1] = 0x00
+	credExistBad.IdenStateData.IdenState[1] ^= 0xff
 	require.NotEqual(t, credExist, credExistBad)
 	err = verifier.VerifyCredentialExistence(credExistBad)
 	assert.NotNil(t, err)

--- a/core/claims/claimAuthorizeKSignBabyJub_test.go
+++ b/core/claims/claimAuthorizeKSignBabyJub_test.go
@@ -23,8 +23,10 @@ func testClaimAuthorizeKSignBabyJub(t *testing.T, i, testKey string) {
 	assert.True(t, merkletree.CheckEntryInField(*c0.Entry()))
 	e := c0.Entry()
 	// Check claim against test vector
-	testgen.CheckTestValue(t, "ClaimAuthorizeKSignBabyJub"+i+"_HIndex", e.HIndex().Hex())
-	testgen.CheckTestValue(t, "ClaimAuthorizeKSignBabyJub"+i+"_HValue", e.HValue().Hex())
+	hi, hv, err := e.HiHv()
+	assert.Nil(t, err)
+	testgen.CheckTestValue(t, "ClaimAuthorizeKSignBabyJub"+i+"_HIndex", hi.Hex())
+	testgen.CheckTestValue(t, "ClaimAuthorizeKSignBabyJub"+i+"_HValue", hv.Hex())
 	testgen.CheckTestValue(t, "ClaimAuthorizeKSignBabyJub"+i+"_dataString", e.Data.String())
 	dataTestOutput(&e.Data)
 	c1 := NewClaimAuthorizeKSignBabyJubFromEntry(e)

--- a/core/claims/claimBasic_test.go
+++ b/core/claims/claimBasic_test.go
@@ -23,8 +23,10 @@ func TestClaimBasic(t *testing.T) {
 	c0.Metadata().RevNonce = 5678
 	e := c0.Entry()
 	// Check claim against test vector
-	testgen.CheckTestValue(t, "ClaimBasic0_HIndex", e.HIndex().Hex())
-	testgen.CheckTestValue(t, "ClaimBasic0_HValue", e.HValue().Hex())
+	hi, hv, err := e.HiHv()
+	assert.Nil(t, err)
+	testgen.CheckTestValue(t, "ClaimBasic0_HIndex", hi.Hex())
+	testgen.CheckTestValue(t, "ClaimBasic0_HValue", hv.Hex())
 	testgen.CheckTestValue(t, "ClaimBasic0_dataString", e.Data.String())
 	dataTestOutput(&e.Data)
 	c1 := NewClaimBasicFromEntry(e)
@@ -48,9 +50,11 @@ func TestClaimBasic1(t *testing.T) {
 	// ClaimBasic
 	c0 := NewClaimBasic(indexSlot, valueSlot)
 	e := c0.Entry()
+	hi, hv, err := e.HiHv()
+	assert.Nil(t, err)
 	// Check claim against test vector
-	testgen.CheckTestValue(t, "ClaimBasic1_HIndex", e.HIndex().Hex())
-	testgen.CheckTestValue(t, "ClaimBasic1_HValue", e.HValue().Hex())
+	testgen.CheckTestValue(t, "ClaimBasic1_HIndex", hi.Hex())
+	testgen.CheckTestValue(t, "ClaimBasic1_HValue", hv.Hex())
 	testgen.CheckTestValue(t, "ClaimBasic1_dataString", e.Data.String())
 	dataTestOutput(&e.Data)
 	c1 := NewClaimBasicFromEntry(e)

--- a/core/claims/revocationrootsleafs_test.go
+++ b/core/claims/revocationrootsleafs_test.go
@@ -16,8 +16,10 @@ func TestLeafRootsTree(t *testing.T) {
 	l0 := NewLeafRootsTree(root)
 	e := l0.Entry()
 
-	testgen.CheckTestValue(t, "Leaf0_HIndex", e.HIndex().Hex())
-	testgen.CheckTestValue(t, "Leaf0_HValue", e.HValue().Hex())
+	hi, hv, err := e.HiHv()
+	assert.Nil(t, err)
+	testgen.CheckTestValue(t, "Leaf0_HIndex", hi.Hex())
+	testgen.CheckTestValue(t, "Leaf0_HValue", hv.Hex())
 	testgen.CheckTestValue(t, "Leaf0_dataString", e.Data.String())
 	l1 := NewLeafRootsTreeFromEntry(e)
 	assert.Equal(t, l0, l1)
@@ -33,8 +35,10 @@ func TestLeafRevocationsTree(t *testing.T) {
 	l0 := NewLeafRevocationsTree(nonce, version)
 	e := l0.Entry()
 
-	testgen.CheckTestValue(t, "Leaf1_HIndex", e.HIndex().Hex())
-	testgen.CheckTestValue(t, "Leaf1_HValue", e.HValue().Hex())
+	hi, hv, err := e.HiHv()
+	assert.Nil(t, err)
+	testgen.CheckTestValue(t, "Leaf1_HIndex", hi.Hex())
+	testgen.CheckTestValue(t, "Leaf1_HValue", hv.Hex())
 	testgen.CheckTestValue(t, "Leaf1_dataString", e.Data.String())
 	l1 := NewLeafRevocationsTreeFromEntry(e)
 	assert.Equal(t, l0, l1)
@@ -55,7 +59,9 @@ func TestAddLeafRootsTree(t *testing.T) {
 	assert.Nil(t, err)
 	testgen.CheckTestValue(t, "rootRootsTree0", mt.RootKey().Hex())
 
-	proof, err := mt.GenerateProof(NewLeafRootsTree(root).Entry().HIndex(), nil)
+	hi, err := NewLeafRootsTree(root).Entry().HIndex()
+	assert.Nil(t, err)
+	proof, err := mt.GenerateProof(hi, nil)
 	assert.Nil(t, err)
 	testgen.CheckTestValue(t, "proofLeafRootsTree", hex.EncodeToString(proof.Bytes()))
 }
@@ -71,7 +77,9 @@ func TestAddLeafRevocationsTree(t *testing.T) {
 	assert.Nil(t, err)
 	testgen.CheckTestValue(t, "rootRevocationsTree0", mt.RootKey().Hex())
 
-	proof, err := mt.GenerateProof(NewLeafRevocationsTree(nonce, version).Entry().HIndex(), nil)
+	hi, err := NewLeafRevocationsTree(nonce, version).Entry().HIndex()
+	assert.Nil(t, err)
+	proof, err := mt.GenerateProof(hi, nil)
 	assert.Nil(t, err)
 	testgen.CheckTestValue(t, "proofRevocationsTree", hex.EncodeToString(proof.Bytes()))
 }

--- a/identity/holder/holder.go
+++ b/identity/holder/holder.go
@@ -77,7 +77,11 @@ func (h *Holder) HolderGetCredentialValidity(
 	claimMetadata.Unmarshal(credExist.Claim)
 	// NOTE: Once we add versions, this will require some changes that need to be thought properly!
 	revLeaf := claims.NewLeafRevocationsTree(claimMetadata.RevNonce, 0xffffffff).Entry()
-	mtpNotNonce, err := publicData.RevocationsTree.GenerateProof(revLeaf.HIndex(), nil)
+	revLeafHi, err := revLeaf.HIndex()
+	if err != nil {
+		return nil, err
+	}
+	mtpNotNonce, err := publicData.RevocationsTree.GenerateProof(revLeafHi, nil)
 	if err != nil {
 		return nil, err
 	}

--- a/identity/issuer/issuer.go
+++ b/identity/issuer/issuer.go
@@ -622,7 +622,12 @@ func (is *Issuer) RevokeClaim(claim merkletree.Entrier) error {
 	}
 	is.rw.Lock()
 	defer is.rw.Unlock()
-	data, err := is.claimsTree.GetDataByIndex(claim.Entry().HIndex())
+
+	hi, err := claim.Entry().HIndex()
+	if err != nil {
+		return err
+	}
+	data, err := is.claimsTree.GetDataByIndex(hi)
 	if err != nil {
 		return err
 	}
@@ -711,7 +716,11 @@ func (is *Issuer) GenCredentialExistence(claim merkletree.Entrier) (*proof.Crede
 		return nil, err
 	}
 	claimEntry := claim.Entry()
-	mtpExist, err := generateExistenceMTProof(is.claimsTree, claimEntry.HIndex(),
+	hi, err := claimEntry.HIndex()
+	if err != nil {
+		return nil, err
+	}
+	mtpExist, err := generateExistenceMTProof(is.claimsTree, hi,
 		idenStateTreeRoots.ClaimsTreeRoot)
 	if err != nil {
 		// We were unable to generate a proof from the claims tree

--- a/merkletree/merkletree.go
+++ b/merkletree/merkletree.go
@@ -145,21 +145,38 @@ func (e *Entry) Value() []ElemBytes {
 	return e.Data[IndexLen:]
 }
 
-// HIndex calculates the hash of the Index of the entry, used to find the path
+// HIndex calculates the hash of the Index of the Entry, used to find the path
 // from the root to the leaf in the MT.
-func (e *Entry) HIndex() *Hash {
+func (e *Entry) HIndex() (*Hash, error) {
+	var err error
 	if e.hIndex == nil { // Cache the hIndex.
 		//e.hIndex = HashElems(e.Index()[:]...)
-		e.hIndex = HashElems(e.Index()...)
+		e.hIndex, err = HashElems(e.Index()...)
 	}
-	return e.hIndex
+	return e.hIndex, err
 }
 
-func (e *Entry) HValue() *Hash {
+// HValue calculates the hash of the Value of the Entry
+func (e *Entry) HValue() (*Hash, error) {
+	var err error
 	if e.hValue == nil { // Cache the hValue.
-		e.hValue = HashElems(e.Value()...)
+		e.hValue, err = HashElems(e.Value()...)
 	}
-	return e.hValue
+	return e.hValue, err
+}
+
+// HiHv returns the HIndex and HValue of the Entry
+func (e *Entry) HiHv() (*Hash, *Hash, error) {
+	hi, err := e.HIndex()
+	if err != nil {
+		return nil, nil, err
+	}
+	hv, err := e.HValue()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return hi, hv, nil
 }
 
 func (e *Entry) Bytes() []byte {
@@ -207,7 +224,10 @@ func NewMerkleTree(storage db.Storage, maxLevels int) (*MerkleTree, error) {
 			return nil, err
 		}
 		nodeRoot := NewNodeEmpty()
-		k, _ := nodeRoot.Key(), nodeRoot.Value()
+		k, err := nodeRoot.Key()
+		if err != nil {
+			return nil, err
+		}
 		mt.rootKey = k
 		mt.dbInsert(tx, rootNodeValue, DBEntryTypeRoot, mt.rootKey[:])
 		if err = tx.Commit(); err != nil {
@@ -262,7 +282,11 @@ func (mt *MerkleTree) GetDataByIndex(hIndex *Hash) (*Data, error) {
 		case NodeTypeEmpty:
 			return nil, ErrEntryIndexNotFound
 		case NodeTypeLeaf:
-			if bytes.Equal(hIndex[:], n.Entry.HIndex()[:]) {
+			hi, err := n.Entry.HIndex()
+			if err != nil {
+				return nil, err
+			}
+			if bytes.Equal(hIndex[:], hi[:]) {
 				return &n.Entry.Data, nil
 			} else {
 				return nil, ErrEntryIndexNotFound
@@ -290,7 +314,11 @@ func (mt *MerkleTree) EntryExists(entry *Entry, rootKey *Hash) error {
 			return err
 		}
 	}
-	data, err := mt.GetDataByIndex(entry.HIndex())
+	hi, err := entry.HIndex()
+	if err != nil {
+		return err
+	}
+	data, err := mt.GetDataByIndex(hi)
 	if err != nil {
 		return err
 	}
@@ -322,13 +350,22 @@ func (mt *MerkleTree) pushLeaf(tx db.Tx, newLeaf *Node, oldLeaf *Node,
 		}
 		return mt.addNode(tx, newNodeMiddle)
 	} else {
+		oldLeafKey, err := oldLeaf.Key()
+		if err != nil {
+			return nil, err
+		}
+		newLeafKey, err := newLeaf.Key()
+		if err != nil {
+			return nil, err
+		}
+
 		if pathNewLeaf[lvl] {
-			newNodeMiddle = NewNodeMiddle(oldLeaf.Key(), newLeaf.Key())
+			newNodeMiddle = NewNodeMiddle(oldLeafKey, newLeafKey)
 		} else {
-			newNodeMiddle = NewNodeMiddle(newLeaf.Key(), oldLeaf.Key())
+			newNodeMiddle = NewNodeMiddle(newLeafKey, oldLeafKey)
 		}
 		// We can add newLeaf now.  We don't need to add oldLeaf because it's already in the tree.
-		_, err := mt.addNode(tx, newLeaf)
+		_, err = mt.addNode(tx, newLeaf)
 		if err != nil {
 			return nil, err
 		}
@@ -354,9 +391,16 @@ func (mt *MerkleTree) addLeaf(tx db.Tx, newLeaf *Node, key *Hash,
 		return mt.addNode(tx, newLeaf)
 	case NodeTypeLeaf:
 		// TODO: delete old node n???  Make this optional???
-		hIndex := n.Entry.HIndex()
+		hIndex, err := n.Entry.HIndex()
+		if err != nil {
+			return nil, err
+		}
 		// Check if leaf node found contains the leaf node we are trying to add
-		if bytes.Equal(hIndex[:], newLeaf.Entry.HIndex()[:]) {
+		newLeafHi, err := newLeaf.Entry.HIndex()
+		if err != nil {
+			return nil, err
+		}
+		if bytes.Equal(hIndex[:], newLeafHi[:]) {
 			return nil, ErrEntryIndexAlreadyExists
 		}
 		pathOldLeaf := getPath(mt.maxLevels, hIndex)
@@ -421,7 +465,10 @@ func (mt *MerkleTree) AddEntry(e *Entry) error {
 	}()
 
 	newNodeLeaf := NewNodeLeaf(e)
-	hIndex := e.HIndex()
+	hIndex, err := e.HIndex()
+	if err != nil {
+		return err
+	}
 	path := getPath(mt.maxLevels, hIndex)
 
 	newRootKey, err := mt.addLeaf(tx, newNodeLeaf, mt.rootKey, 0, path)
@@ -478,11 +525,16 @@ func (mt *MerkleTree) GraphViz(w io.Writer, rootKey *Hash) error {
 node [fontname=Monospace,fontsize=10,shape=box]
 `)
 	cnt := 0
+	var errIn error
 	err := mt.Walk(rootKey, func(n *Node) {
+		k, err := n.Key()
+		if err != nil {
+			errIn = err
+		}
 		switch n.Type {
 		case NodeTypeEmpty:
 		case NodeTypeLeaf:
-			fmt.Fprintf(w, "\"%v\" [style=filled];\n", n.Key())
+			fmt.Fprintf(w, "\"%v\" [style=filled];\n", k)
 		case NodeTypeMiddle:
 			lr := [2]string{n.ChildL.String(), n.ChildR.String()}
 			for i := range lr {
@@ -492,11 +544,14 @@ node [fontname=Monospace,fontsize=10,shape=box]
 					cnt++
 				}
 			}
-			fmt.Fprintf(w, "\"%v\" -> {\"%v\" \"%v\"}\n", n.Key(), lr[0], lr[1])
+			fmt.Fprintf(w, "\"%v\" -> {\"%v\" \"%v\"}\n", k, lr[0], lr[1])
 		default:
 		}
 	})
 	fmt.Fprintf(w, "}\n")
+	if errIn != nil {
+		return errIn
+	}
 	return err
 }
 
@@ -513,7 +568,11 @@ func (mt *MerkleTree) DumpTree(w io.Writer, rootKey *Hash) error {
 	var errS error
 	err := mt.Walk(rootKey, func(n *Node) {
 		if n.Type != NodeTypeEmpty {
-			err := serializeKV(w, n.Key().Bytes(), n.Value())
+			k, err := n.Key()
+			if err != nil {
+				errS = err
+			}
+			err = serializeKV(w, k.Bytes(), n.Value())
 			if err != nil {
 				errS = err
 			}
@@ -832,12 +891,16 @@ func (mt *MerkleTree) GenerateProof(hIndex *Hash, rootKey *Hash) (*Proof, error)
 		case NodeTypeEmpty:
 			return p, nil
 		case NodeTypeLeaf:
-			if bytes.Equal(hIndex[:], n.Entry.HIndex()[:]) {
+			nHi, nHv, err := n.Entry.HiHv()
+			if err != nil {
+				return nil, err
+			}
+			if bytes.Equal(hIndex[:], nHi[:]) {
 				p.Existence = true
 				return p, nil
 			} else {
 				// We found a leaf whose entry didn't match hIndex
-				p.nodeAux = &nodeAux{hIndex: n.Entry.HIndex(), hValue: n.Entry.HValue()}
+				p.nodeAux = &nodeAux{hIndex: nHi, hValue: nHv}
 				return p, nil
 			}
 		case NodeTypeMiddle:
@@ -873,9 +936,13 @@ func VerifyProof(rootKey *Hash, proof *Proof, hIndex, hValue *Hash) bool {
 // hValue.
 func RootFromProof(proof *Proof, hIndex, hValue *Hash) (*Hash, error) {
 	sibIdx := len(proof.Siblings) - 1
+	var err error
 	var midKey *Hash
 	if proof.Existence {
-		midKey = LeafKey(hIndex, hValue)
+		midKey, err = LeafKey(hIndex, hValue)
+		if err != nil {
+			return nil, err
+		}
 	} else {
 		if proof.nodeAux == nil {
 			midKey = &HashZero
@@ -883,7 +950,10 @@ func RootFromProof(proof *Proof, hIndex, hValue *Hash) (*Hash, error) {
 			if bytes.Equal(hIndex[:], proof.nodeAux.hIndex[:]) {
 				return nil, fmt.Errorf("Non-existence proof being checked against hIndex equal to nodeAux")
 			}
-			midKey = LeafKey(proof.nodeAux.hIndex, proof.nodeAux.hValue)
+			midKey, err = LeafKey(proof.nodeAux.hIndex, proof.nodeAux.hValue)
+			if err != nil {
+				return nil, err
+			}
 		}
 	}
 	path := getPath(int(proof.depth), hIndex)
@@ -896,9 +966,15 @@ func RootFromProof(proof *Proof, hIndex, hValue *Hash) (*Hash, error) {
 			siblingKey = &HashZero
 		}
 		if path[lvl] {
-			midKey = NewNodeMiddle(siblingKey, midKey).Key()
+			midKey, err = NewNodeMiddle(siblingKey, midKey).Key()
+			if err != nil {
+				return nil, err
+			}
 		} else {
-			midKey = NewNodeMiddle(midKey, siblingKey).Key()
+			midKey, err = NewNodeMiddle(midKey, siblingKey).Key()
+			if err != nil {
+				return nil, err
+			}
 		}
 	}
 	return midKey, nil
@@ -925,9 +1001,13 @@ func (mt *MerkleTree) addNode(tx db.Tx, n *Node) (*Hash, error) {
 		return nil, ErrNotWritable
 	}
 	if n.Type == NodeTypeEmpty {
-		return n.Key(), nil
+		return n.Key()
 	}
-	k, v := n.Key(), n.Value()
+	k, err := n.Key()
+	if err != nil {
+		return nil, err
+	}
+	v := n.Value()
 	// Check that the node key doesn't already exist
 	if _, err := tx.Get(k[:]); err == nil {
 		return nil, ErrNodeKeyAlreadyExists

--- a/merkletree/merkletree_test.go
+++ b/merkletree/merkletree_test.go
@@ -83,7 +83,9 @@ func TestEntry(t *testing.T) {
 	// in := testgen.GetTestValue("EntryInts0").([]int64)
 	in := interfaceToInt64Array(testgen.GetTestValue("EntryInts0"))
 	e := NewEntryFromIntArray(in)
-	testgen.CheckTestValue(t, "TestEntry0", hex.EncodeToString(e.HIndex()[:]))
+	hi, err := e.HIndex()
+	assert.Nil(t, err)
+	testgen.CheckTestValue(t, "TestEntry0", hex.EncodeToString(hi[:]))
 }
 
 func TestData(t *testing.T) {
@@ -163,16 +165,24 @@ func TestEntriesIndex(t *testing.T) {
 	// Two entries with different Index generate different hash index
 	in := interfaceToInt64Array(testgen.GetTestValue("EntryInts4"))
 	a := NewEntryFromIntArray(in)
+	aHi, err := a.HIndex()
+	assert.Nil(t, err)
 	in = interfaceToInt64Array(testgen.GetTestValue("EntryInts5"))
 	b := NewEntryFromIntArray(in)
-	assert.NotEqual(t, a.HIndex(), b.HIndex())
+	bHi, err := b.HIndex()
+	assert.Nil(t, err)
+	assert.NotEqual(t, aHi, bHi)
 
 	// Two entries with same Index generate the same hash index
 	in = interfaceToInt64Array(testgen.GetTestValue("EntryInts6"))
 	c := NewEntryFromIntArray(in)
+	cHi, err := c.HIndex()
+	assert.Nil(t, err)
 	in = interfaceToInt64Array(testgen.GetTestValue("EntryInts7"))
 	d := NewEntryFromIntArray(in)
-	assert.Equal(t, c.HIndex(), d.HIndex())
+	dHi, err := d.HIndex()
+	assert.Nil(t, err)
+	assert.Equal(t, cHi, dHi)
 }
 
 func TestGetEntry2(t *testing.T) {
@@ -186,11 +196,13 @@ func TestGetEntry2(t *testing.T) {
 	}
 	in = interfaceToInt64Array(testgen.GetTestValue("EntryInts1"))
 	e = NewEntryFromIntArray(in)
+	hi, err := e.HIndex()
+	assert.Nil(t, err)
 	if err := mt.AddEntry(&e); err != nil {
 		t.Fatal(err)
 	}
 
-	data, err := mt.GetDataByIndex(e.HIndex())
+	data, err := mt.GetDataByIndex(hi)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -205,11 +217,13 @@ func TestGenerateProof1(t *testing.T) {
 
 	in := interfaceToInt64Array(testgen.GetTestValue("EntryInts8"))
 	e := NewEntryFromIntArray(in)
+	hi, err := e.HIndex()
+	assert.Nil(t, err)
 	if err := mt.AddEntry(&e); err != nil {
 		t.Fatal(err)
 	}
 
-	proof, err := mt.GenerateProof(e.HIndex(), nil)
+	proof, err := mt.GenerateProof(hi, nil)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -228,8 +242,10 @@ func TestGenerateProof4(t *testing.T) {
 	}
 
 	e := NewEntryFromInts(int64(2), 0, 0, 0, 0, 0, 0, 0)
+	hi, err := e.HIndex()
+	assert.Nil(t, err)
 
-	data, err := mt.GetDataByIndex(e.HIndex())
+	data, err := mt.GetDataByIndex(hi)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -237,7 +253,7 @@ func TestGenerateProof4(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	proof, err := mt.GenerateProof(e.HIndex(), nil)
+	proof, err := mt.GenerateProof(hi, nil)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -257,8 +273,10 @@ func TestGenerateProof64(t *testing.T) {
 	}
 
 	e := NewEntryFromInts(int64(4), 0, 0, 0, 0, 0, 0, 0)
+	hi, err := e.HIndex()
+	assert.Nil(t, err)
 
-	data, err := mt.GetDataByIndex(e.HIndex())
+	data, err := mt.GetDataByIndex(hi)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -266,7 +284,7 @@ func TestGenerateProof64(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	proof, err := mt.GenerateProof(e.HIndex(), nil)
+	proof, err := mt.GenerateProof(hi, nil)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -286,13 +304,15 @@ func TestVerifyProof1(t *testing.T) {
 	}
 
 	e := NewEntryFromInts(int64(4), 0, 0, 0, 0, 0, 0, 0)
+	hi, hv, err := e.HiHv()
+	assert.Nil(t, err)
 
-	proof, err := mt.GenerateProof(e.HIndex(), nil)
+	proof, err := mt.GenerateProof(hi, nil)
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	verify := VerifyProof(mt.RootKey(), proof, e.HIndex(), e.HValue())
+	verify := VerifyProof(mt.RootKey(), proof, hi, hv)
 	assert.True(t, verify)
 	proofTestOutput(proof)
 	testgen.CheckTestValue(t, "TestVerifyProof1", hex.EncodeToString(proof.Bytes()))
@@ -310,13 +330,15 @@ func TestVerifyProofEmpty(t *testing.T) {
 	}
 
 	e := NewEntryFromInts(int64(42), 0, 0, 0, 0, 0, 0, 0)
+	hi, hv, err := e.HiHv()
+	assert.Nil(t, err)
 
-	proof, err := mt.GenerateProof(e.HIndex(), nil)
+	proof, err := mt.GenerateProof(hi, nil)
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	verify := VerifyProof(mt.RootKey(), proof, e.HIndex(), e.HValue())
+	verify := VerifyProof(mt.RootKey(), proof, hi, hv)
 	assert.True(t, verify)
 	proofTestOutput(proof)
 	testgen.CheckTestValue(t, "TestVerifyProofEmpty", hex.EncodeToString(proof.Bytes()))
@@ -335,18 +357,23 @@ func TestVerifyProofCases(t *testing.T) {
 
 	// Existence proof
 	e := NewEntryFromInts(int64(4), 0, 0, 0, 0, 0, 0, 0)
-	proof, err := mt.GenerateProof(e.HIndex(), nil)
+	hi, hv, err := e.HiHv()
+	assert.Nil(t, err)
+
+	proof, err := mt.GenerateProof(hi, nil)
 	if err != nil {
 		t.Fatal(err)
 	}
 	assert.Equal(t, proof.Existence, true)
-	assert.True(t, VerifyProof(mt.RootKey(), proof, e.HIndex(), e.HValue()))
+	assert.True(t, VerifyProof(mt.RootKey(), proof, hi, hv))
 	proofTestOutput(proof)
 	testgen.CheckTestValue(t, "TestVerifyProofCases0", hex.EncodeToString(proof.Bytes()))
 
 	for i := 8; i < 32; i++ {
 		e := NewEntryFromInts(int64(i), 0, 0, 0, 0, 0, 0, 0)
-		proof, err = mt.GenerateProof(e.HIndex(), nil)
+		hi, err := e.HIndex()
+		assert.Nil(t, err)
+		proof, err = mt.GenerateProof(hi, nil)
 		assert.Nil(t, err)
 		if debug {
 			fmt.Println(i, proof)
@@ -354,25 +381,29 @@ func TestVerifyProofCases(t *testing.T) {
 	}
 	// Non-existence proof, empty aux
 	e = NewEntryFromInts(int64(12), 0, 0, 0, 0, 0, 0, 0)
-	proof, err = mt.GenerateProof(e.HIndex(), nil)
+	hi, hv, err = e.HiHv()
+	assert.Nil(t, err)
+	proof, err = mt.GenerateProof(hi, nil)
 	if err != nil {
 		t.Fatal(err)
 	}
 	assert.Equal(t, proof.Existence, false)
 	// assert.True(t, proof.nodeAux == nil)
-	assert.True(t, VerifyProof(mt.RootKey(), proof, e.HIndex(), e.HValue()))
+	assert.True(t, VerifyProof(mt.RootKey(), proof, hi, hv))
 	proofTestOutput(proof)
 	testgen.CheckTestValue(t, "TestVerifyProofCases1", hex.EncodeToString(proof.Bytes()))
 
 	// Non-existence proof, diff. node aux
 	e = NewEntryFromInts(int64(10), 0, 0, 0, 0, 0, 0, 0)
-	proof, err = mt.GenerateProof(e.HIndex(), nil)
+	hi, hv, err = e.HiHv()
+	assert.Nil(t, err)
+	proof, err = mt.GenerateProof(hi, nil)
 	if err != nil {
 		t.Fatal(err)
 	}
 	assert.Equal(t, proof.Existence, false)
 	// assert.True(t, proof.nodeAux != nil)
-	assert.True(t, VerifyProof(mt.RootKey(), proof, e.HIndex(), e.HValue()))
+	assert.True(t, VerifyProof(mt.RootKey(), proof, hi, hv))
 	proofTestOutput(proof)
 	testgen.CheckTestValue(t, "TestVerifyProofCases2", hex.EncodeToString(proof.Bytes()))
 }
@@ -391,17 +422,23 @@ func TestVerifyProofFalse(t *testing.T) {
 	// Invalid existence proof (node used for verification doesn't
 	// correspond to node in the proof)
 	e := NewEntryFromInts(int64(4), 0, 0, 0, 0, 0, 0, 0)
-	proof, err := mt.GenerateProof(e.HIndex(), nil)
+	hi, err := e.HIndex()
+	assert.Nil(t, err)
+	proof, err := mt.GenerateProof(hi, nil)
 	if err != nil {
 		t.Fatal(err)
 	}
 	assert.Equal(t, proof.Existence, true)
 	e1 := NewEntryFromInts(int64(5), 0, 0, 0, int64(5), 0, 0, 0)
-	assert.True(t, !VerifyProof(mt.RootKey(), proof, e1.HIndex(), e1.HValue()))
+	hi, hv, err := e1.HiHv()
+	assert.Nil(t, err)
+	assert.True(t, !VerifyProof(mt.RootKey(), proof, hi, hv))
 
 	// Invalid non-existence proof (Non-existence proof, diff. node aux)
 	e = NewEntryFromInts(int64(4), 0, 0, 0, 0, 0, 0, 0)
-	proof, err = mt.GenerateProof(e.HIndex(), nil)
+	hi, err = e.HIndex()
+	assert.Nil(t, err)
+	proof, err = mt.GenerateProof(hi, nil)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -409,8 +446,10 @@ func TestVerifyProofFalse(t *testing.T) {
 	// Now we change the proof from existence to non-existence, and add e's
 	// data as auxiliary node.
 	proof.Existence = false
-	proof.nodeAux = &nodeAux{hIndex: e.HIndex(), hValue: e.HValue()}
-	assert.True(t, !VerifyProof(mt.RootKey(), proof, e.HIndex(), e.HValue()))
+	hi, hv, err = e.HiHv()
+	assert.Nil(t, err)
+	proof.nodeAux = &nodeAux{hIndex: hi, hValue: hv}
+	assert.True(t, !VerifyProof(mt.RootKey(), proof, hi, hv))
 }
 
 func TestProofFromBytesSmall(t *testing.T) {
@@ -423,7 +462,9 @@ func TestProofFromBytesSmall(t *testing.T) {
 	}
 
 	// Proof of existence, single claim MT
-	proof0, err := mt.GenerateProof(e0.HIndex(), nil)
+	hi, err := e0.HIndex()
+	assert.Nil(t, err)
+	proof0, err := mt.GenerateProof(hi, nil)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -433,7 +474,9 @@ func TestProofFromBytesSmall(t *testing.T) {
 
 	// Proof of non-existence with aux node, single claim MT
 	e2 := NewEntryFromInts(int64(1), 0, 0, 0, 0, 0, 0, 0)
-	proof2, err := mt.GenerateProof(e2.HIndex(), nil)
+	hi, err = e2.HIndex()
+	assert.Nil(t, err)
+	proof2, err := mt.GenerateProof(hi, nil)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -455,7 +498,9 @@ func TestProofFromBytesBig(t *testing.T) {
 
 	// Proof of existence, single claim MT
 	e0 := NewEntryFromInts(0, 0, 0, 0, 0, 0, 0, 0)
-	proof0, err := mt.GenerateProof(e0.HIndex(), nil)
+	hi, err := e0.HIndex()
+	assert.Nil(t, err)
+	proof0, err := mt.GenerateProof(hi, nil)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -465,7 +510,9 @@ func TestProofFromBytesBig(t *testing.T) {
 
 	// Proof of non-existence with empty node, single claim MT
 	e1 := NewEntryFromInts(int64(17), 0, 0, 0, 0, 0, 0, 0)
-	proof1, err := mt.GenerateProof(e1.HIndex(), nil)
+	hi, err = e1.HIndex()
+	assert.Nil(t, err)
+	proof1, err := mt.GenerateProof(hi, nil)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -475,7 +522,9 @@ func TestProofFromBytesBig(t *testing.T) {
 
 	// Proof of non-existence with aux node, single claim MT
 	e2 := NewEntryFromInts(0, int64(17), 0, 0, 0, 0, 0, 0)
-	proof2, err := mt.GenerateProof(e2.HIndex(), nil)
+	hi, err = e2.HIndex()
+	assert.Nil(t, err)
+	proof2, err := mt.GenerateProof(hi, nil)
 	if err != nil {
 		t.Fatal(err)
 	}

--- a/merkletree/node.go
+++ b/merkletree/node.go
@@ -83,7 +83,7 @@ func NewNodeFromBytes(b []byte) (*Node, error) {
 
 // LeafKey computes the key of a leaf node given the hIndex and hValue of the
 // entry of the leaf.
-func LeafKey(hIndex, hValue *Hash) *Hash {
+func LeafKey(hIndex, hValue *Hash) (*Hash, error) {
 	// return HashElems(ElemBytesOne, ElemBytes(*hIndex), ElemBytes(*hValue))
 	return HashElemsKey(big.NewInt(1), ElemBytes(*hIndex), ElemBytes(*hValue))
 }
@@ -91,21 +91,32 @@ func LeafKey(hIndex, hValue *Hash) *Hash {
 // Key computes the key of the node by hashing the content in a specific way
 // for each type of node.  This key is used as the hash of the merklee tree for
 // each node.
-func (n *Node) Key() *Hash {
+func (n *Node) Key() (*Hash, error) {
 	if n.key == nil { // Cache the key to avoid repeated hash computations.
 		// NOTE: We are not using the type to calculate the hash!
 		switch n.Type {
 		case NodeTypeMiddle: // H(ChildL || ChildR)
-			n.key = HashElems(ElemBytes(*n.ChildL), ElemBytes(*n.ChildR))
+			var err error
+			n.key, err = HashElems(ElemBytes(*n.ChildL), ElemBytes(*n.ChildR))
+			if err != nil {
+				return nil, err
+			}
 		case NodeTypeLeaf: // H(Data...)
-			n.key = LeafKey(n.Entry.HIndex(), n.Entry.HValue())
+			hi, hv, err := n.Entry.HiHv()
+			if err != nil {
+				return nil, err
+			}
+			n.key, err = LeafKey(hi, hv)
+			if err != nil {
+				return nil, err
+			}
 		case NodeTypeEmpty: // Zero
 			n.key = &HashZero
 		default:
 			n.key = &HashZero
 		}
 	}
-	return n.key
+	return n.key, nil
 }
 
 // Value returns the value of the node.  This is the content that is stored in the backend database.

--- a/merkletree/utils.go
+++ b/merkletree/utils.go
@@ -63,20 +63,20 @@ func BigIntToHash(e *big.Int) (h Hash) {
 	return h
 }
 
-// HashElems performs a mimc7 hash over the array of ElemBytes.
-func HashElems(elems ...ElemBytes) *Hash {
+// HashElems performs a poseidon hash over the array of ElemBytes.
+func HashElems(elems ...ElemBytes) (*Hash, error) {
 	bigints := ElemBytesToBigInts(elems...)
 	// mimcHash, err := mimc7.Hash(bigints, nil)
 	poseidonHash, err := poseidon.Hash(bigints)
 	if err != nil {
-		panic(err)
+		return nil, err
 	}
 	h := BigIntToHash(poseidonHash)
-	return &h
+	return &h, nil
 }
 
-// HashElemsKey performs a mimc7 hash over the array of ElemBytes.
-func HashElemsKey(key *big.Int, elems ...ElemBytes) *Hash {
+// HashElemsKey performs a poseidon hash over the array of ElemBytes.
+func HashElemsKey(key *big.Int, elems ...ElemBytes) (*Hash, error) {
 	bigints := ElemBytesToBigInts(elems...)
 	// mimcHash, err := mimc7.Hash(bigints, key)
 	if key != nil {
@@ -84,10 +84,10 @@ func HashElemsKey(key *big.Int, elems ...ElemBytes) *Hash {
 	}
 	poseidonHash, err := poseidon.Hash(bigints)
 	if err != nil {
-		panic(err)
+		return nil, err
 	}
 	h := BigIntToHash(poseidonHash)
-	return &h
+	return &h, nil
 }
 
 // getPath returns the binary path, from the root to the leaf.

--- a/merkletree/utils_test.go
+++ b/merkletree/utils_test.go
@@ -29,25 +29,29 @@ func TestGetSetBitmap(t *testing.T) {
 func TestHashElems(t *testing.T) {
 	in := interfaceToInt64Array(testgen.GetTestValue("TestHashElems0"))
 	d := IntsToData(in[0], in[1], in[2], in[3], in[4], in[5], in[6], in[7])
-	h := HashElems(d[:]...)
+	h, err := HashElems(d[:]...)
+	assert.Nil(t, err)
 	hashTestOutput(h)
 	testgen.CheckTestValue(t, "TestHashElems0", h.Hex())
 
 	in = interfaceToInt64Array(testgen.GetTestValue("TestHashElems1"))
 	d = IntsToData(in[0], in[1], in[2], in[3], in[4], in[5], in[6], in[7])
-	h = HashElems(d[:]...)
+	h, err = HashElems(d[:]...)
+	assert.Nil(t, err)
 	hashTestOutput(h)
 	testgen.CheckTestValue(t, "TestHashElems1", h.Hex())
 
 	in = interfaceToInt64Array(testgen.GetTestValue("TestHashElems2"))
 	d = IntsToData(in[0], in[1], in[2], in[3], in[4], in[5], in[6], in[7])
-	h = HashElems(d[:]...)
+	h, err = HashElems(d[:]...)
+	assert.Nil(t, err)
 	hashTestOutput(h)
 	testgen.CheckTestValue(t, "TestHashElems2", h.Hex())
 
 	in = interfaceToInt64Array(testgen.GetTestValue("EntryInts0"))
 	d = IntsToData(in[0], in[1], in[2], in[3], in[4], in[5], in[6], in[7])
-	h = HashElems(d[:]...)
+	h, err = HashElems(d[:]...)
+	assert.Nil(t, err)
 	hashTestOutput(h)
 	testgen.CheckTestValue(t, "TestHashElems3", h.Hex())
 }

